# Продолжаю улучшать свои навыки и знания. 

**Предлагаю ознакомиться с "мини проектом Bookings", в котором привожу решение типовых задач.**

## Задачи

* Выяснить пользователи из каких стран совершили наибольшее число успешных бронирований.
* Выбрать страны, входящие в топ-5 по количеству бронирований. 
* Выяснить среднее количество бронирований ночей в отелях разных типов.

* Выяснить сколько раз происходил овербукинг. В данном примере овербукинг, это изменения тип номера, полученного клиентом,  от изначально забронированного. 
* Выяснить на какой месяц чаще всего успешно оформляли бронь в 2016? Изменился ли самый популярный месяц в 2017.
* Выяснить на какой месяц бронирования отеля типа City Hotel отменялись чаще всего в 2015, 2016 и 2017 годах. 

* Выяснить "Double Occupancy" — среднее количество проживающих в номере, для каждого типа отелей необходимо узнать число взрослых, детей и младенцев. 
* Выяснить для какого типа отелей, максимальное среднее число клиентов с детьми и младенцами оказалось наибольшим?

* Выяснить отношение количество посетителей отменивших бронирование к общему количеству клиентов, результат предоставить в процентах. Churn rate (коэффициент оттока).
* Выяснить, в какой группе посетителей, показатель отмененных бронирований выше. Например: посетители с детьми или посетители без детей больше совершали отмену бронирования.



Имеются следующие переменные:

* Hotel – тип отеля (City Hotel или Resort Hotel)
* Is canceled – бронирование было отменено (1) или нет (0); не отмененное считается успешным
* Lead time – количество дней, прошедших между датой бронирования и датой прибытия
* Arrival full date – полная дата прибытия
* Arrival date year – год прибытия
* Arrival date month – месяц прибытия
* Arrival date week number – номер недели прибытия
* Arrival date day of month – день прибытия
* Stays in weekend nights – количество выходных (суббота или воскресенье), которые гость забронировал для проживания в  отеле
* Stays in week nights – количество дней (с понедельника по пятницу), которые гость забронировал для проживания в отеле
* Stays total nights – общее число забронированных ночей (сумма двух предыдущих колонок)
* Adults – число взрослых
* Children – число детей
* Babies – число младенцев
* Meal – выбранный тип питания
* Country – страна происхождения клиента
* Reserved room type – тип зарезервированного номера
* Assigned room type – тип полученного номера (может отличаться от забронированного)
* Customer type – тип бронирования
* Reservation status – значение последнего статуса брони: Canceled - было отменено клиентом; Check-Out - клиент зарегистрировался, но уже покинул отель; No-Show - клиент не зарегистрировался и сообщил администрации отеля причину
* Reservation status date – дата обновления статуса


In [1]:
import pandas as pd

In [2]:
bookings = pd.read_csv('https://stepik.org/media/attachments/course/111949/2_bookings.csv',  sep=';')

In [3]:
bookings_head = bookings.head(7)

In [4]:
bookings.shape

(119390, 21)

In [5]:
bookings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 21 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   Hotel                      119390 non-null  object 
 1   Is Canceled                119390 non-null  int64  
 2   Lead Time                  119390 non-null  int64  
 3   arrival full date          119390 non-null  object 
 4   Arrival Date Year          119390 non-null  int64  
 5   Arrival Date Month         119390 non-null  object 
 6   Arrival Date Week Number   119390 non-null  int64  
 7   Arrival Date Day of Month  119390 non-null  int64  
 8   Stays in Weekend nights    119390 non-null  int64  
 9   Stays in week nights       119390 non-null  int64  
 10  stays total nights         119390 non-null  int64  
 11  Adults                     119390 non-null  int64  
 12  Children                   119386 non-null  float64
 13  Babies                     11

In [6]:
bookings.dtypes

Hotel                         object
Is Canceled                    int64
Lead Time                      int64
arrival full date             object
Arrival Date Year              int64
Arrival Date Month            object
Arrival Date Week Number       int64
Arrival Date Day of Month      int64
Stays in Weekend nights        int64
Stays in week nights           int64
stays total nights             int64
Adults                         int64
Children                     float64
Babies                         int64
Meal                          object
Country                       object
Reserved Room Type            object
Assigned room type            object
customer type                 object
Reservation Status            object
Reservation status_date       object
dtype: object

In [7]:
bookings.dtypes.value_counts()

object     10
int64      10
float64     1
dtype: int64

In [8]:
bookings.head(9)

Hotel  Is Canceled  Lead Time arrival full date  Arrival Date Year  \
0  Resort Hotel            0        342        2015-07-01               2015   
1  Resort Hotel            0        737        2015-07-01               2015   
2  Resort Hotel            0          7        2015-07-01               2015   
3  Resort Hotel            0         13        2015-07-01               2015   
4  Resort Hotel            0         14        2015-07-01               2015   
5  Resort Hotel            0         14        2015-07-01               2015   
6  Resort Hotel            0          0        2015-07-01               2015   
7  Resort Hotel            0          9        2015-07-01               2015   
8  Resort Hotel            1         85        2015-07-01               2015   

  Arrival Date Month  Arrival Date Week Number  Arrival Date Day of Month  \
0               July                        27                          1   
1               July                        27                          1   
2               July                        27                          1   
3               July                        27                          1   
4               July                        27                          1   
5               July                        27                          1   
6               July                        27                          1   
7               July                        27                          1   
8               July                        27                          1   

   Stays in Weekend nights  Stays in week nights  ...  Adults  Children  \
0                        0                     0  ...       2       0.0   
1                        0                     0  ...       2       0.0   
2                        0                     1  ...       1       0.0   
3                        0                     1  ...       1       0.0   
4                        0                     2  ...       2       0.0   
5                        0                     2  ...       2       0.0   
6                        0                     2  ...       2       0.0   
7                        0                     2  ...       2       0.0   
8                        0                     3  ...       2       0.0   

   Babies  Meal Country Reserved Room Type Assigned room type customer type  \
0       0    BB     PRT                  C                  C     Transient   
1       0    BB     PRT                  C                  C     Transient   
2       0    BB     GBR                  A                  C     Transient   
3       0    BB     GBR                  A                  A     Transient   
4       0    BB     GBR                  A                  A     Transient   
5       0    BB     GBR                  A                  A     Transient   
6       0    BB     PRT                  C                  C     Transient   
7       0    FB     PRT                  C                  C     Transient   
8       0    BB     PRT                  A                  A     Transient   

  Reservation Status Reservation status_date  
0          Check-Out              2015-07-01  
1          Check-Out              2015-07-01  
2          Check-Out              2015-07-02  
3          Check-Out              2015-07-02  
4          Check-Out              2015-07-03  
5          Check-Out              2015-07-03  
6          Check-Out              2015-07-03  
7          Check-Out              2015-07-03  
8           Canceled              2015-05-06  

[9 rows x 21 columns]

Как вы уже могли заметить, при сборе датасета произошла небольшая ошибка, и названия колонок были записаны в неудобном для дальнейшей работы формате. Исправлю ситуацию, приведя названия столбцов к нижнему регистру и заменив пробелы на знак нижнего подчеркивания. 
Например: Is Canceled -> is_canceled
 * lower() - переобразует буквы из заглавных в строчные
 * replace(' ', '_') - заменит пробелы на анднрскоры

In [9]:
'Arrival Date Year'.lower()

'arrival date year'

In [10]:
'Arrival Date Year'.replace(' ', '_')

'Arrival_Date_Year'

In [11]:
def to_lower_undercore(name):
    name = name.lower().replace(' ', '_')
    return name

In [12]:
to_lower_undercore('Arrival Date Year') #proverka ok

'arrival_date_year'

In [13]:
bookings = bookings.rename(columns=to_lower_undercore)

In [14]:
bookings

hotel  is_canceled  lead_time arrival_full_date  \
0       Resort Hotel            0        342        2015-07-01   
1       Resort Hotel            0        737        2015-07-01   
2       Resort Hotel            0          7        2015-07-01   
3       Resort Hotel            0         13        2015-07-01   
4       Resort Hotel            0         14        2015-07-01   
...              ...          ...        ...               ...   
119385    City Hotel            0         23        2017-08-30   
119386    City Hotel            0        102        2017-08-31   
119387    City Hotel            0         34        2017-08-31   
119388    City Hotel            0        109        2017-08-31   
119389    City Hotel            0        205        2017-08-29   

        arrival_date_year arrival_date_month  arrival_date_week_number  \
0                    2015               July                        27   
1                    2015               July                        27   
2                    2015               July                        27   
3                    2015               July                        27   
4                    2015               July                        27   
...                   ...                ...                       ...   
119385               2017             August                        35   
119386               2017             August                        35   
119387               2017             August                        35   
119388               2017             August                        35   
119389               2017             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
0                               1                        0   
1                               1                        0   
2                               1                        0   
3                               1                        0   
4                               1                        0   
...                           ...                      ...   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  ...  adults  children  babies  meal country  \
0                          0  ...       2       0.0       0    BB     PRT   
1                          0  ...       2       0.0       0    BB     PRT   
2                          1  ...       1       0.0       0    BB     GBR   
3                          1  ...       1       0.0       0    BB     GBR   
4                          2  ...       2       0.0       0    BB     GBR   
...                      ...  ...     ...       ...     ...   ...     ...   
119385                     5  ...       2       0.0       0    BB     BEL   
119386                     5  ...       3       0.0       0    BB     FRA   
119387                     5  ...       2       0.0       0    BB     DEU   
119388                     5  ...       2       0.0       0    BB     GBR   
119389                     7  ...       2       0.0       0    HB     DEU   

       reserved_room_type assigned_room_type customer_type reservation_status  \
0                       C                  C     Transient          Check-Out   
1                       C                  C     Transient          Check-Out   
2                       A                  C     Transient          Check-Out   
3                       A                  A     Transient          Check-Out   
4                       A                  A     Transient          Check-Out   
...                   ...                ...           ...                ...   
119385                  A                  A     Transient          Check-Out   
119386         

Перейду к исследованию данных. Выясню:
* Пользователи из каких стран совершили наибольшее число успешных бронирований;
* Бронирование считается успешным, если в дальнейшем не было отменено (переменная is_canceled). 
* В качестве ответа выберу страны, входящие в топ-5. сохраню в переменную top_5_countries

In [15]:
top_5_countries = bookings.query('is_canceled == 0') \
        .country \
        .value_counts()[:5]

top_5_countries

# [:5] выводит 5 наибольших

PRT    21071
GBR     9676
FRA     8481
ESP     6391
DEU     6069
Name: country, dtype: int64

Узнаю на сколько ночей (stays_total_nights)  в среднем бронируют отели типа City Hotel? Resort Hotel? Полученные значения округлю с точностью до 2 знаков после точки.

In [16]:
bookings.groupby('hotel') \
        .agg({ 'stays_total_nights' : 'mean'}) \
        .round(2)
 

# mean - функция среднего значения
# raund(2) - округление до 2 х знаков после точки

stays_total_nights
hotel                           
City Hotel                  2.98
Resort Hotel                4.32

Иногда тип номера, полученного клиентом (assigned_room_type), отличается от изначально забронированного (reserved_room_type). Такое может произойти, например, по причине овербукинга.

Узнаю сколько подобных наблюдений встретилось в датасете?

In [17]:
bookings[['reserved_room_type', 'assigned_room_type', 'hotel']] 
# для просмотра более одной колонок, необходимо создавать списки используя [] внутри []

reserved_room_type assigned_room_type         hotel
0                       C                  C  Resort Hotel
1                       C                  C  Resort Hotel
2                       A                  C  Resort Hotel
3                       A                  A  Resort Hotel
4                       A                  A  Resort Hotel
...                   ...                ...           ...
119385                  A                  A    City Hotel
119386                  E                  E    City Hotel
119387                  D                  D    City Hotel
119388                  A                  A    City Hotel
119389                  A                  A    City Hotel

[119390 rows x 3 columns]

In [18]:
overbooking = bookings.query('assigned_room_type != reserved_room_type') \
                      .shape[0]

overbooking

# расчитал общий овердбукинг 

# !=  не равны 
# [0] расчет числа строк по нулевому индексу

14917

Узнаю овердбукин в отелях типа "Resort Hotel"  (отели располеженные в прибрежном районе)

In [19]:
overbooking_Resort = bookings.query('assigned_room_type != reserved_room_type and hotel == "Resort Hotel"')\
                    .shape[0]
                                           
overbooking_Resort

7725

Узнаю овердбукин в отелях типа "Resort Hotel" (отели располеженные в центральной части города)

In [20]:
overbooking_City = bookings.query('assigned_room_type != reserved_room_type and hotel != "Resort Hotel"')\
                .shape[0]
                      
                          
                          
overbooking_City

7192

Теперь проанализирую даты запланированного прибытия (arrival_date_year).
Узнаю:
* На какой месяц чаще всего оформляли бронь в 2016 году? 
* Изменился ли самый популярный месяц в 2017?

In [21]:
most_popular_month_2016 = bookings.query('arrival_date_year == 2016') \
                                        .arrival_date_month \
                                        .value_counts() \
                                        .idxmax()
most_popular_month_2016

'October'

In [22]:
most_popular_month_2017 = bookings.query('arrival_date_year == 2017') \
                                    .arrival_date_month \
                                    .value_counts() \
                                    .idxmax()
most_popular_month_2017

'May'

Сгруппирую  данные по годам, а затем проверю, на какой месяц (arrival_date_month) бронирования отеля типа City Hotel отменялись чаще всего в 2015? 2016? 2017? 

In [23]:
(
    bookings
        .query('hotel == "City Hotel" and is_canceled == 1 ')
        .groupby('arrival_date_year')
        .arrival_date_month
        .value_counts()
)

arrival_date_year  arrival_date_month
2015               September             1543
                   October               1321
                   August                1232
                   July                   939
                   December               668
                   November               301
2016               October               1947
                   June                  1720
                   September             1567
                   April                 1539
                   May                   1436
                   November              1360
                   August                1247
                   March                 1108
                   December              1072
                   July                  1043
                   February               930
                   January                438
2017               May                   2217
                   April                 1926
                   June                  1

Посмотрю на числовые характеристики трёх колонок: adults, children и babies. Узнаю какая из них имеет наибольшее среднее значение.

In [24]:
bookings[['adults', 'children', 'babies']].describe()
# describe - выводит: count = строки ,mean = среднее, std = стандарное отклонение(среднеквадратичное), мин и квартиль и максимум

adults       children         babies
count  119390.000000  119386.000000  119390.000000
mean        1.856403       0.103890       0.007949
std         0.579261       0.398561       0.097436
min         0.000000       0.000000       0.000000
25%         2.000000       0.000000       0.000000
50%         2.000000       0.000000       0.000000
75%         2.000000       0.000000       0.000000
max        55.000000      10.000000      10.000000

In [25]:
# еще один способ 

bookings[['adults', 'children', 'babies']].mean()

# mean считаем средние значение

adults      1.856403
children    0.103890
babies      0.007949
dtype: float64

По условию задачи неоходимо выяснить наибольшее среднее значение взрослых, детей и младенцев для каждого типа отелей.

### Double Occupancy — Cреднее количество проживающих в номере

**Double Occupancy** ******показывает, сколько в среднем человек заселяется в один номер. Показатель обычно выше в семейных загородных отелях и ниже в городских отелях, работающих с бизнес-путешественниками или командированными.***** 

In [26]:
double_occupancy_Resort = bookings.query('hotel == "Resort Hotel"')\
[['adults', 'children', 'babies']].mean()

double_occupancy_Resort

adults      1.867149
children    0.128682
babies      0.013904
dtype: float64

In [27]:
double_occupancy_City = bookings.query('hotel == "City Hotel"')\
[['adults', 'children', 'babies']].mean()

double_occupancy_City

adults      1.850977
children    0.091370
babies      0.004941
dtype: float64

**Выясню для какого типа отелей, максимальное среднее число клиентов с детьми и младенцами оказалось наибольшим?**
* Создам колонку total_kids, объеденив столбцы children и babies.
 
* В качестве ответа укажу наибольшее среднее total_kids, округлив до 2 знаков после точки.

In [28]:
bookings['total_kids'] = bookings['children'] + bookings['babies']

In [31]:
total_max_mean_kids = bookings.groupby('hotel') \
                        .agg({'total_kids' : 'mean'}) \
                        .round(2) 
                        
total_max_mean_kids

total_kids
hotel                   
City Hotel          0.10
Resort Hotel        0.14

**Выясню отношение количество посетителей отменивших бронирование к общему количеству клиентов, результат предоставить в процентах. Churn rate (коэффициент оттока).**


#### Churn rate (коэффициент оттока) – это процент клиентов (например, клиенты с детьми), которые отменили бронирование, отказались от услуг сервиса в течение определенного периода времени. Иными словами, представляет собой отношение количества ушедших пользователей к общему количеству пользователей, выраженное в процентах.

* В данном случае Churn Rate - это процент клиентов, которые отменили бронирование.


In [67]:
churn_rate = bookings.query('is_canceled == 1').shape[0] \
                / bookings.query('is_canceled == 0').shape[0]\

round(churn_rate, 2)

0.59

**Выясню, в какой группе посетителей, показатель отмененных бронирований выше, посетители с детьми или посетители без детей больше совершали отмену бронирования.**
* Создам переменную has_kids, которая принимает значение True, если клиент при бронировании указал хотя бы одного ребенка (total_kids), в противном случае – False. Далее проверьте, среди какой группы пользователей показатель оттока выше.

* В качестве ответа укажу наибольший % оттока, округленный до 2 знаков после точки



In [32]:
bookings['has_kids'] = bookings.total_kids > 0
# клиенты у которох есть дети 
# .shape[0]  # Количество пользователей равно количеству строк после сортировки с заданными критерием

In [64]:
not_kids_churn = bookings.query('is_canceled == 1 and has_kids == False').shape[0] \
                / bookings.query('has_kids == False').shape[0]\

round(not_kids_churn, 2)

0.37

In [66]:
yas_kids_churn = bookings.query('is_canceled == 1 and has_kids == True').shape[0] \
                / bookings.query('has_kids == True').shape[0] \

round(yas_kids_churn , 2) 

0.35

Из полученных данных мы можем видеть Churn rate (коэффициент оттока)  у клиентов с детьми 0,35% , а у клиентов без детей 0,37%

# Аналитическая записка

### Итоги аналитического исследования пользователей совершавших бронирование номеров в отелях двух типов.

* В результате подведения аналитического исследования данных 119 389 пользователей, совершавших бронирования в отелях двух типов, мною было выявлены топ-5 стран с максимальным количеством бронирований, а именно 21 071 бронирований совершенно гражданами Португалии, 9 676 бронирований гражданами Соединенного Королевства Великобритании, 8 481 бронирование совершили граждане Франции, 6 391 бронирование совершили граждане Испании, 6069 бронирований совершили граждане Германии.
* Среднее количество бронирований в отели типа City Hotel = 2.98 ночей, в отели типа Resort Hotel = 4.32 ночей.
* Орвербукинг в отелях всех типов 14 917, в отелях типа Resort Hotel = 7 725, в отелях типа City Hotel = города 7 192 .
* Выяснил, что в 2016 году максимальное число бронирований пришлось на май, а в 2017 году на октябрь.
* Проанализировал данные о бронировании отелей типа "City Hotel" за 2015, 2016 и 2017 года, и выявил что в 2015 максимальное количество бронирований 1 543 в сентябре, в 2016 году максимальное количество бронирований 1 947 в октября и в 2017 году максимальное количество бронирований 2 217 в мае.
* Показатель "Double Occupancy" — среднее количество проживающих в номере, для отелей "City Hotel", взрослых = 1.850977, детей =  0.091370 и младенцев = 0.004941. "Double Occupancy", для отелей "Resort Hotel", взрослых = 1.867149, детей = 0.128682 и младенцев = 0.013904.
* Выявил коэффициент оттока (Churn Rate) за весь период данного исследования = 0,59 это процент пользователей, которые отменили бронирование.  
* Исследовал Churn rate (коэффициент оттока), выяснил что, процент отмены бронирования у клиентов с детьми  0,35% , у клиентов без детей 0,37% . 


                                                                                                            С уважением,
                                                                                                            Осокин Максим